### PR TITLE
Migrate widgets to 2.x

### PR DESCRIPTION
**Pull request by unkcpz**
- [x] `computer_resource` widget
- [x] Drop dep aiida-core <=1.x
- [x] Remove all deprecate widgets
- [ ] ~~bump AWB version to `2.0.0`~~
- [x] computer/code setup widget support `aiida-code-registry` for both 1.x and 2.x core datatype.
- [x] Add tests to test computer/code setup from code registry failed for incompatibility.

Blocking the issue: https://github.com/aiidalab/aiidalab-qe/issues/279

If we implement this change means it is not compatible with 1.x anymore. 
I think we need to set more strict dependency for aiida-core and bump the version of this PR to `2.0.0`. 

@yakutovicha @csadorf, any thoughts?

### CODE REVIEW

**Comment by csadorf** [comment]
> If we implement this change means it is not compatible with 1.x anymore. I think we need to set more strict dependency for aiida-core and bump the version of this PR to `2.0.0`.
> 
> @yakutovicha @csadorf, any thoughts?

Yes, if this change requires aiida-core >= 2.0.0 then this needs to be reflected in the declared dependencies.

**Comment by yakutovicha** [comment]
> @yakutovicha @csadorf, any thoughts?

if those changes are enough, I am very happy :)

**Comment by unkcpz** [comment]
We are going to deprecate some widgets such as `computer.py::SshComputerSetup`, I will also remove them.
But always smoke tests are very useful for such changes, is that possible to have the test framework ready for this PR? @csadorf. Let me know how to add it and I'll add the tests.

**Comment by unkcpz** [comment]
> if those changes are enough, I am very happy :)

Sadly, I this may not enough, we need to have a way to separate [aiida-code-registry](https://github.com/aiidateam/aiida-code-registry) to support both 1.x and 2.x config files.

**Comment by unkcpz** [comment]
Tests are all passed. But the computer/code setup configs from https://github.com/aiidateam/aiida-code-registry not working anymore. 
One possible way is we also release `aiida-core-registry` with a tag for instance `1.x` and make the config database support select data by the tag.

**Comment by csadorf** [comment]
> Tests are all passed. But the computer/code setup configs from https://github.com/aiidateam/aiida-code-registry not working anymore. One possible way is we also release `aiida-core-registry` with a tag for instance `1.x` and make the config database support select data by the tag. Do you have any better way? @csadorf @yakutovicha

Can you elaborate on what exactly is not working? Maybe it is possible to do an automated migration? I'd prefer that over maintaining multiple versions for different "target AiiDA versions".

**Comment by unkcpz** [comment]
> Can you elaborate on what exactly is not working?

From aiida 1.x to 2.x the transport and scheduler are changed from an endpoint, for example, `ssh` to `core.ssh` and `slurm` to `core.slurm`. They are hard coded in the config file.

**Comment by yakutovicha** [comment]
> This is for sure a good option but will increase the complexity of the `aiida-code-registry` which is already not very convenient to add a new entity for HPC resources.

Ok, let me tell you my perspective on this.

1. If the registry is not convenient, we should make it convenient. What are the issues with adding a new code/computer there? I am happy to add a code/computer template or anything to make it more user-friendly.
Can you put some suggestions to the registry in a form of an issue? If things are not working as expected, we (as developers) should not be silent about them.
2. I believe that @csadorf is talking about an automated migration within AWB. The problem is - at some point the registry will have to be also migrated, which makes it unavoidable to keep old and new configs there. Tbh, I lean a bit towards the suggestion of @unkcpz. We might automatically generate the computer/code config files that are suited for the given version of AiiDA. It won't mean maintaining two registries, it would rather mean having one registry that automatically generates config files for AiiDA-1.x and AiiDA-2.x. If we agree, I can take care of this approach.

**Comment by unkcpz** [comment]
> If the registry is not convenient, we should make it convenient. What are the issues with adding a new code/computer there? I am happy to add a code/computer template or anything to make it more user-friendly.

No issue with adding new code/computer. I misunderstood that @csadorf ask to have for same configs files have two versions of aiida supported. 
But I have some ideas in mind on how to improve the `aiida-code-registry`, I will make an AEP for it.

> I believe that @csadorf is talking about an automated migration within AWB. 

In this case, we need to think about which version to keep in `main` branch and doing transfer in ABW from 1.x -> 2.x or vice versa. So I'd vote to auto-generate config files AiiDA-1.x and AiiDA-2.x. 

@csadorf Sorry that I didn't read Sasha's comment carefully and did not make it very clear in the materials cloud meeting.

**Comment by csadorf** [comment]
As I explained in the MC meeting today, we need to interpret the code registry as a database that is providing information for the automated configuration of AiiDA codes. The specific semantics (`ssh` vs `core.ssh`) are in that sense completely irrelevant to the database. As long as the information needed to configure codes in aiida-core 1.x and 2.x remains the same we can always do an automated translation between the different semantics, there is no need to maintain two different versions of the same database.

That translation can happen at various points (at the database level itself), when we read from the database, when we actually configure the codes. Is there any kind of API regarding the registry? Are there any other consumers despite the AWB library? If not, then I would suggest to simply migrate the whole database, otherwise, just do a translation here in the library code. In either case there is no need and we should not maintain multiple database versions.

**Comment by danielhollas** [comment]
Just a note, it would be great if you could release one more AWB version before this is merged. :-)

**Comment by danielhollas** [comment]
> I believe you prefer to have https://github.com/aiidalab/aiidalab-widgets-base/pull/348 merged before the release, right?

Yeah, that would be nice. More importantly for me, I am also working on a follow-up on #339, to support multiple structure upload in StructureUploadWidget. I'll submit a PR today and then you can judge whether it can be merged quickly or will need more thought and should not block release. Thanks!

**Comment by unkcpz** [comment]
Hi @yakutovicha, since this is not block the aiidalab-qe, I think it is ready to have another round of review.

File: aiidalab_widgets_base/computational_resources.py
Comment (csadorf) [inline aiidalab_widgets_base/computational_resources.py:709]: I don't think it's a good idea to ask the user to provide the value in kB since the number of machines providing RAM on the order of kB should at this point be restricted to edge and embedded device that are not typically known to run AiiDA or Jupyter. You can always internally convert the value if that is what is expected by the code plugin.

File: aiidalab_widgets_base/computational_resources.py
Comment (csadorf) [inline aiidalab_widgets_base/computational_resources.py:909]: Where is this default value coming from?

File: aiidalab_widgets_base/databases.py
Comment (csadorf) [inline aiidalab_widgets_base/databases.py:288]: I am not sure that the `hold_trait_notification()` context manager is really appropriate here. Are we changing multiple traits at once within the same thread? The hold should not be confused with a lock. See the [docs](https://traitlets.readthedocs.io/en/stable/using_traitlets.html#holding-trait-cross-validation-and-notifications) for reference.

Also, why not make the request outside of the hold?

_Since my comment is not directly related to this change set, please address it in a separate issue or PR if any action is deemed necessary._

File: aiidalab_widgets_base/process.py
Comment (csadorf) [inline aiidalab_widgets_base/process.py:100]: I'd recommend the following sequence:
```python
if callable(inputs_generator):
    self.inputs_generator = inputs_generator
else:
    raise ValueError("The `inputs_generator` argument must be provided and a function that ...")
```

File: .readthedocs.yml
Comment (csadorf) [inline .readthedocs.yml:13]: Why not 3.10?

**Review by csadorf — DISMISSED** [review]

File: aiidalab_widgets_base/computational_resources.py
Comment (unkcpz) [reply aiidalab_widgets_base/computational_resources.py:909]: It's should be set for AiiDA 2.x, I think @yakutovicha add it to the computer setup and made it mandatory.
It is hard to say if we should provide the default value for this field, maybe we shouldn't provide the default one. 
But this comes with the problem that the code-registry-database doesn't have this field set yet, the "quick setup" of widget will fail since no value is provided.

File: aiidalab_widgets_base/process.py
Comment (unkcpz) [reply aiidalab_widgets_base/process.py:100]: I don't understand what you mean, the code you paste here is all the same.

Comment (csadorf) [reply aiidalab_widgets_base/process.py:100]: You have an additional `is None` check which is unnecessary I think (`None` is not `callable`).

Comment (unkcpz) [reply aiidalab_widgets_base/process.py:96]: Thanks! I see. I remove the `None` as the default value of `inputs_generator` argument.

File: aiidalab_widgets_base/computational_resources.py
Comment (csadorf) [reply aiidalab_widgets_base/computational_resources.py:909]: I don't really think it matters where it comes from, but I think there should be some kind of justification for it and then it should be documented here in the form of a short comment.

**Review by yakutovicha — COMMENTED** [review]
Thanks, @unkcpz. Not too many comments from my side.

**Comment by unkcpz** [comment]
@yakutovicha thanks!

> There is a way to have no default value for the memory when running verdi computer setup. It is !, which corresponds to the None as an internal value.

The `self.default_memory_per_machine` is an `ipw.IntText` where the value should be an int. I think there is no simple to use the `None` for the value, it must be an int initial number and allow the user to increase or decrease the number. 
One possible solution is using Text type and accept None as input.

**Comment by unkcpz** [comment]
I use `-1` for not setting the default value and works fine in my test.

**Comment by unkcpz** [comment]
Hi @yakutovicha, I still don't quite understand what you mean of we can set the `default_memory_per_machine` as `None`. Yes, when it pass to ComputerBuilder it can be `None` then the value just not set. 
But here the problem is more about we have the widget now I call it `default_memory_per_machine_gb` of type `ipw.IntText`, and it must has an int value. If I didn't set the value for the widget, the value will be `0` as default. 
Please have a look at my current implementation, if the value is `0` I regard it as not set.

**Comment by unkcpz** [comment]
I think there is a more subtle issue with the aiida 2.x which will break the computer setup widget here. 
When a new computer setup from the `ComputerBuilder` API, a new user will be created that bind to the computer even if the email is the same, and this user cannot be shown by `verdi user list`. 
If the computer setup from the command line interface it will use the `default_user` for the computer being setup.

Pining @sphuber @chrisjsewell for input. Can you explain a bit about how the user creates and bind to the computer?

**Comment by sphuber** [comment]
A `Computer` doesn't have a user. It is the `AuthInfo` class that has a `User` and is joined on a `Computer`. So when you run `verdi computer setup` a `Computer` instance is created that has no associated user. It is when you run `verdi computer configure` (which does `Computer.configure()` ) that an `AuthInfo` is created for the computer for a particular `User`.

**Comment by unkcpz** [comment]
Thanks @sphuber! It works.

@yakutovicha, I fix the default user issue and rebase the PR, now the newly added code can all be show in the drowdown. Please make sure to rebase and add more commits to it.

**Comment by danielhollas** [comment]
Hi @unkcpz,

I've been testing my app with the new `aiidalab/full-stack` images and this AWB branch, and I've ran into the same issue as you describe in https://github.com/aiidalab/aiidalab-qe/issues/279 with `'not persistent within this Session'.

As you noted elsewhere, it seems the problem is caused by multithreading. Have you tested various widgets here that use multithreading?

I am reusing some code from aiidalab-qe, and the error came from the `CalcJobOutputFollower`, https://github.com/aiidalab/aiidalab-qe/blob/379d20977c0e6fb85c8494fdc8f78cdcecb6ebc5/aiidalab_qe/widgets.py#L243.
I can see similar code in AWB widget `ProcessMonitor` in `aiidalab_widgets_base/process.py` so was wondering if widgets in that file work.

**Comment by unkcpz** [comment]
Hi @danielhollas, sorry I just see the message somehow I didn't get GitHub notifications today. 
I think you are correct, the problem happened when two thread access aiida database. If the tread locks are implemented properly, it shouldn't be a problem. I'll take a close look at all these cases.

**Comment by danielhollas** [comment]
Hi,

Yes, CalcJobOutputFollower is a straight copy form QeApp. I don't know how the current QeApp looks, but you should be able to trigger it if you click on a CalcJob node in a NodeTreeWidget. Let me know if you need more detailed info and I'll get to you when I am at my computer.

**Comment by unkcpz** [comment]
Hi @danielhollas @yakutovicha, I think I have a clear view of how/why AiiDA 2.x is incompatible with the thread widgets. From my point of view, the key to solving it is to prevent using traitslet object being the aiida object gets from other threads. I think it has to be using Node UUID for those traitlets, I don't think there is another solution for it.

Can we have a meeting to settle it and I will work on changing those widgets first.

**Comment by unkcpz** [comment]
I create a minimal example for it, using a list as the mutable object to mimic traitlets. 

```python
from aiida import orm
from aiida import load_profile
import threading

load_profile()

def _db_operation(mut: list):
    code = orm.load_code("doubler@localhost")
    mut.append(code)

def main():
    """Query the list of available codes."""

    mut = []
    thread = threading.Thread(target=_db_operation, args=(mut,))
    thread.start()
    thread.join()
    
    code = mut[0]
    print(code)

    
if __name__ == "__main__":
    main()
```

End up with `Instance '<DbNode at 0x7f714adcfdc0>' is not persistent within this Session`.

**Comment by danielhollas** [comment]
Hi @unkcpz,

I am assuming that the refactor to use UUIDs for traitlets would be only needed for widgets that use threading? 

Thank you for the example, that's helpful. It would be great to see an example of a refactored widget to better understand the implications.

**Comment by unkcpz** [comment]
> I am assuming that the refactor to use UUIDs for traitlets would be only needed for widgets that use threading?

Yes and no. The widget may not use thread but its methods can be triggered by other thread and update aiida object traitlets from that thread and cause the issue. It is more subtle to find and prevent.

But I think refactoring the widget using treading is the first step. And since we must use threading in those cases, changing the traitlets by using UUID is the only solution I think. 
If you and @yakutovicha have no objection, I'll rebase this PR and add a refactoring to those widgets.

**Comment by danielhollas** [comment]
@unkcpz given the already long discussion here and many other changes, I would recommend landing this PR first, maybe to a temporary branch 'aiida2' for example, and then open a new PR with the threading changes against it.

**Review by yakutovicha — COMMENTED** [review]
I have a little comment. Otherwise, good to go.

We should probably merge this and start releasing an alpha version of 2.0.0 to keep going.

The remaining tasks from a to-do list should probably go to separate issues.

**Review by danielhollas — APPROVED** [review]

File: .github/workflows/ci.yml
Comment (danielhollas) [inline .github/workflows/ci.yml:31]: The tag is not used in this file anymore.

File: .github/workflows/ci.yml
Comment (danielhollas) [inline .github/workflows/ci.yml:46]: Maybe the Python version should be defined globally since it is at two places now.

File: aiidalab_widgets_base/computational_resources.py
Comment (danielhollas) [inline aiidalab_widgets_base/computational_resources.py:772]: Maybe add a tooltip with a more detailed explanation of what this is used for? Also the comment in the code should explain in more detail why the escaping is needed.

File: aiidalab_widgets_base/computational_resources.py
Comment (danielhollas) [inline aiidalab_widgets_base/computational_resources.py:959]: I don't understand why the 1024 value is squared?

File: tests/docker-compose.yml
Comment (danielhollas) [inline tests/docker-compose.yml:18]: Any reason not to use `8888:8888`?

**Review by danielhollas — CHANGES_REQUESTED** [review]
@unkcpz thanks so much for the tests, that's awesome that you've been able to make it work. :tada: 

Just a few minor comments / questions from me.

Since this is basically ready and has the tests, I would suggest to merge this one first, and build the uuid PR on top of it. I propose to have a temporary branch (`aiida2`) where we merge this and then we can thoroughly test on the uuid PR before merging to master. What do you think @yakutovicha

File: tests/docker-compose.yml
Comment (unkcpz) [reply tests/docker-compose.yml:7]: You are correct, the latest is the "stable" one and should be used for other apps. `edge` is point to the latest commit.

File: .github/workflows/ci.yml
Comment (unkcpz) [reply .github/workflows/ci.yml:31]: I keep it since going to use it as an ENV_VAR for using in the docker-compose. I'll change the docker-compose file.

File: tests/docker-compose.yml
Comment (unkcpz) [reply tests/docker-compose.yml:18]: Ha nice catch. Only personal reason since the `8888` of my workstation is occupied by some wired raid service. Will change it back to `8888`.

**Comment by yakutovicha** [comment]
> Since this is basically ready and has the tests, I would suggest to merge this one first, and build the uuid PR on top of it. I propose to have a temporary branch (`aiida2`) where we merge this and then we can thoroughly test on the uuid PR before merging to master. What do you think @yakutovicha

I would merge those things directly to `master` and make only bug fixes in the [support/aiida-1.x](https://github.com/aiidalab/aiidalab-widgets-base/tree/support/aiida-1.x) branch. I would release 1.4.x versions from the support/aiida-1.x branch.

**Comment by unkcpz** [comment]
> I would merge those things directly to master and make only bug fixes in the [support/aiida-1.x](https://github.com/aiidalab/aiidalab-widgets-base/tree/support/aiida-1.x) branch. I would release 1.4.x versions from the support/aiida-1.x branch.

I agree! 
The issue I mentioned for Node viewer is not dependent on the traitlets UUID change PR. I will fix the things mentioned in @danielhollas review and then should be good to go.

File: aiidalab_widgets_base/computational_resources.py
Comment (yakutovicha) [reply aiidalab_widgets_base/computational_resources.py:959]: The value in the database is in kB, but we agreed previously to keep GB for the widget. Just to answer why 1024 is squared.

I also noticed that this line is wrong, though. Thanks, @danielhollas, for pointing at it. The values shouldn't be an integer but a string. 

Possible solutions are:
```python
self.default_memory_per_machine_widget.value =
         str(decimal.Decimal(f"{value/1024**2:0.3f}").normalize()) + ' GB'
```
This requires importing `decimal`.

Another solution is simply to say that the imported value is in kB
```python
f"{value KB}
```

Any approach is good for me.

I am going to disappear, for some time, so I let @unkcpz and/or @danielhollas pick up an appropriate solution.

File: tests/docker-compose.yml
Comment (unkcpz) [reply tests/docker-compose.yml:18]: Hi @danielhollas, I am thinking maybe it is better to keep the host port other than `8888`, since `8888` is the default port for jupyter and high possibility that already been used by user. This docker-compose.yml will only be used for test purposes and is triggered by `pytest`, it more convenient if it does not conflict with `8888`.

File: aiidalab_widgets_base/computational_resources.py
Comment (unkcpz) [reply aiidalab_widgets_base/computational_resources.py:960]: I prefer the second one without the `decimal` dependency.

**Review by danielhollas — DISMISSED** [review]
Thanks @unkcpz.

I am still not a fan of the current `default_memory_per_machine` widget. To me the fact that it mixes both number and text is super confusing, not to mention that we're introducing an extra dependency just to parse this string. When this field is not prefilled from the loaded computer, it is not clear what the units are, or even that one can specify the units.

Instead I would suggest to use two separate widgets, `ipw.IntText` (or `ipw.BoundedIntText`) for the numerical value itself, and `ipw.Dropdown` for specifying the units, perhaps backed by an Enum like this.

```python
class MemoryUnits(IntEnum):
    KB = 1024
    MB = 1024**2
    GB = 1024**3

self.memory_units = ipw.Dropdown(
    options=[(unit.name, unit) for unit in MemoryUnits],
    value=MemoryUnits.MB,
    description=""
)   
```

Nevertheless, I am approving so this can be done and iterated upon in a separate PR.

**Comment by yakutovicha** [comment]
> I am still not a fan of the current `default_memory_per_machine` widget. To me the fact that it mixes both number and text is super confusing, not to mention that we're introducing an extra dependency just to parse this string. When this field is not prefilled from the loaded computer, it is not clear what the units are, or even that one can specify the units.

Since I am responsible for this one, I try to response.

From my perspective, the fact that the widget is human readable is rather a plus: one can essentially put any value here and make sure that things will work. If one writes something that can’t be parsed, the cross will appear making sure the user is doing something wrong.

if the field is not filled - it is not used. I added a placeholder that tries to explain that. Feel free to suggest a better one.

**Review by yakutovicha — DISMISSED** [review]
Thanks @unkcpz 👍

File: .github/workflows/ci.yml
Comment (unkcpz) [reply .github/workflows/ci.yml]: True, changed. Can you approve it again?

Comment (unkcpz) [reply .github/workflows/ci.yml]: Nevermind, since this is a tiny fix if the tests passed. I can force merge it anyway.

**Review by danielhollas — APPROVED** [review]

**Pull request merged**
